#**PROBLEMA DE NEGOCIO**

Se trata de un algoritmo capas de analizar las caracteristicas de una propiedad -como comodidades, tamaño, ocupación del espacio en un periodo de tiempo determinado- y sugerir al anfitrion un precio a cobrar por tarifas diarias que garantice ganancias en momento de alta demanda.

#**1 entendiendo el problema**

In [ ]:
#IMPORTAR LA LIBRERIA PANDAS
import pandas as pd

In [ ]:
datos = pd.read_json("datos_hosting.json")

In [ ]:
datos.head()

info_inmuebles
0  {'evaluacion_general': '10.0', 'experiencia_lo...
1  {'evaluacion_general': '10.0', 'experiencia_lo...
2  {'evaluacion_general': '10.0', 'experiencia_lo...
3  {'evaluacion_general': '10.0', 'experiencia_lo...
4  {'evaluacion_general': '10.0', 'experiencia_lo...

In [ ]:
#NORMALIZAR EL DATAFRAME

datos = pd.json_normalize(datos["info_inmuebles"])
datos.head()

evaluacion_general experiencia_local max_hospedes  \
0               10.0                --            1   
1               10.0                --           10   
2               10.0                --           11   
3               10.0                --           12   
4               10.0                --           14   

                                   descripcion_local  \
0  [This clean and comfortable one bedroom sits r...   
1  [Welcome to the heart of the 'Ballard Brewery ...   
2  [New modern house built in 2013.  Spectacular ...   
3  [Our NW style home is 3200+ sq ft with 3 level...   
4  [Perfect for groups. 2 bedrooms, full bathroom...   

                                descripcion_vecindad  \
0  [Lower Queen Anne is near the Seattle Center (...   
1  [--, Capital Hill is the heart of Seattle, bor...   
2  [Upper Queen Anne is a charming neighborhood f...   
3  [The Views from our top floor!  Wallingford ha...   
4  [Safeway grocery store within walking distance...   

                                      cantidad_baños  \
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                  [2, 3, 2, 3, 3, 3, 2, 1, 2, 2, 2]   
2                                                [4]   
3                           [3, 3, 3, 3, 3, 3, 3, 3]   
4                                             [2, 3]   

                                    cantidad_cuartos  \
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                  [3, 4, 2, 3, 3, 3, 3, 3, 3, 4, 3]   
2                                                [5]   
3                           [6, 6, 5, 5, 5, 5, 4, 4]   
4                                             [2, 6]   

                                      cantidad_camas  \
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                  [5, 6, 8, 3, 3, 5, 4, 5, 6, 7, 4]   
2                                                [7]   
3                           [6, 6, 7, 8, 7, 7, 6, 6]   
4                                             [3, 9]   

                                         modelo_cama  \
0  [Real Bed, Futon, Futon, Pull-out Sofa, Real B...   
1  [Real Bed, Real Bed, Real Bed, Real Bed, Real ...   
2                                         [Real Bed]   
3  [Real Bed, Real Bed, Real Bed, Real Bed, Real ...   
4                               [Real Bed, Real Bed]   

                                         comodidades  \
0  [{Internet,"Wireless Internet",Kitchen,"Free P...   
1  [{TV,Internet,"Wireless Internet",Kitchen,"Fre...   
2  [{TV,"Cable TV",Internet,"Wireless Internet","...   
3  [{Internet,"Wireless Internet",Kitchen,"Free P...   
4  [{TV,Internet,"Wireless Internet",Kitchen,"Fre...   

                                      cuota_deposito  \
0  [$0, $0, $0, $0, $0, $350.00, $350.00, $350.00...   
1  [$500.00, $300.00, $0, $300.00, $300.00, $360....   
2                                        [$1,000.00]   
3  [$500.00, $500.00, $500.00, $500.00, $500.00, ...   
4                               [$300.00, $2,000.00]   

                                      cuota_limpieza  \
0  [$0, $0, $0, $20.00, $15.00, $28.00, $35.00, $...   
1  [$125.00, $100.00, $85.00, $110.00, $110.00, $...   
2                                          [$300.00]   
3  [$225.00, $300.00, $250.00, $250.00, $250.00, ...   
4                                  [$40.00, $150.00]   

                                              precio  
0  [$110.00, $45.00, $55.00, $52.00, $85.00, $50....  
1  [$350.00, $300.00, $425.00, $300.00, $285.00, ...  
2                                          [$975.00]  
3  [$490.00, $550.00, $350.00, $350.00, $350.00, ...  
4                                 [$200.00, $545.00]

Cuando trabajamos con cualquier conjunto de datos, necesitamos saber qué información nos aportan esos datos, porque solo así podremos estudiarlos y analizarlos para desarrollar una solución de análisis y procesamiento de datos.

En este curso trabajaremos con el conjunto de datos presentes en el archivo datos_hosting.json y, para avanzar en nuestros estudios sobre los datos que proporciona este archivo, entenderemos qué información trae cada columna.

- **evaluacion_general:** se refiere a la puntuación media otorgada para evaluar el alojamiento en la propiedad.
- **experiencia_local:** describe las experiencias ofrecidas durante su estancia en la propiedad.
- **max_hospedes:** informa el número máximo de invitados que permite la ubicación.
- **descripcion_local:** describe la propiedad.
- **descripcion_vecindad:** describe el vecindario alrededor de la propiedad.
- **cantidad_baños:** informa el número de baños disponibles.
- **cantidad_cuartos:** informa el número de habitaciones disponibles.
- **cantidad_camas:** informa el número de camas disponibles.
- **modelo_cama:** informa el modelo de cama ofrecido.
- **comodidades:** informa las comodidades que ofrece la propiedad.
- **cuota_deposito:** informa la tarifa mínima de depósito para la seguridad del hosting.
- **cuota_limpieza:** informa el cargo cobrado por el servicio de limpieza.
precio: se refiere al precio base a cobrar por la estancia diaria en la propiedad.

In [ ]:
#SELECIONAMOS LAS COLUMNAS NECESARIAS QUE CONTIENEN LISTAS DE DATOS PARA EXTRAER

columnas = list(datos.columns)
columnas

['evaluacion_general',
 'experiencia_local',
 'max_hospedes',
 'descripcion_local',
 'descripcion_vecindad',
 'cantidad_baños',
 'cantidad_cuartos',
 'cantidad_camas',
 'modelo_cama',
 'comodidades',
 'cuota_deposito',
 'cuota_limpieza',
 'precio']

In [ ]:
#EXTRAEMOS LAS COLUMNAS QUE CONTIENEN LISTAS A UTILIZAR

datos = datos.explode(columnas[3:])
datos

evaluacion_general experiencia_local max_hospedes  \
0                10.0                --            1   
0                10.0                --            1   
0                10.0                --            1   
0                10.0                --            1   
0                10.0                --            1   
..                ...               ...          ...   
68                nan                --            8   
68                nan                --            8   
68                nan                --            8   
69                nan                --            9   
69                nan                --            9   

                                    descripcion_local  \
0   This clean and comfortable one bedroom sits ri...   
0   Our century old Upper Queen Anne house is loca...   
0   Cozy room in two-bedroom apartment along the l...   
0   Very lovely and cozy room for one. Convenientl...   
0   The “Studio at Mibbett Hollow' is in a Beautif...   
..                                                ...   
68  Beautiful craftsman home in the historic Wedgw...   
68  Located in a very easily accessible area of Se...   
68  This home is fully furnished and available wee...   
69  This business-themed modern home features:  *H...   
69  This welcoming home is in the quiet residentia...   

                                 descripcion_vecindad cantidad_baños  \
0   Lower Queen Anne is near the Seattle Center (s...              1   
0   Upper Queen Anne is a really pleasant, unique ...              1   
0   The convenience of being in Seattle but on the...              1   
0   Ballard is lovely, vibrant and one of the most...              1   
0                                                  --              1   
..                                                ...            ...   
68                                                 --              3   
68  Quiet, dead end street near I-5. The proximity...              2   
68                                                 --              1   
69  Your hosts made Madison Valley their home when...              2   
69  Despite the quiet, leafy streets of this resid...              1   

   cantidad_cuartos cantidad_camas    modelo_cama  \
0                 1              1       Real Bed   
0                 1              1          Futon   
0                 1              1          Futon   
0                 1              1  Pull-out Sofa   
0                 1              1       Real Bed   
..              ...            ...            ...   
68                4              5       Real Bed   
68                4              4       Real Bed   
68                3              4       Real Bed   
69                3              6       Real Bed   
69                4              6       Real Bed   

                                          comodidades cuota_deposito  \
0   {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
0   {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
0   {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
0   {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
0   {"Wireless Internet",Kitchen,"Free Parking on ...             $0   
..                                                ...            ...   
68  {TV,"Cable TV",Internet,"Wireless Internet","A...      $1,000.00   
68  {TV,"Cable TV",Internet,"Wireless Internet",Ki...             $0   
68  {TV,"Cable TV",Internet,"Wireless Internet","A...             $0   
69  {TV,"Cable TV",Internet,"Wireless Internet","A...      $1,000.00   
69  {TV,Internet,"Wireless Internet",Kitchen,"Free...        $500.00   

   cuota_limpieza   precio  
0              $0  $110.00  
0              $0   $45.00  
0              $0   $55.00  
0          $20.00   $52.00  
0          $15.00   $85.00  
..            ...      ...  
68        $178.00  $299.00  
68         $99.00  $199.00  
68             $0

In [ ]:
#CONSULTAMOS LA INFO DEL DATAFRAME PARA MOSTRAR EL TIPO DE DATO DE LAS COLUMNAS
#INDICA QUE TODAS SON OBJECT QUE ENGLOBA CARACTERES EN GENERAL
#DEBERIA INDICAR EL TIPO DE DATO REAL DE CADA COLUMNA COMO LOS TIPO ENTEROS O DECIMALES TAMBIEN
#SE DEBE CORREGIR LOS TIPOS DE DATOS CON LA LIBRERIA NUMPY
datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3818 entries, 0 to 69
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   evaluacion_general    3818 non-null   object
 1   experiencia_local     3818 non-null   object
 2   max_hospedes          3818 non-null   object
 3   descripcion_local     3818 non-null   object
 4   descripcion_vecindad  3818 non-null   object
 5   cantidad_baños        3818 non-null   object
 6   cantidad_cuartos      3818 non-null   object
 7   cantidad_camas        3818 non-null   object
 8   modelo_cama           3818 non-null   object
 9   comodidades           3818 non-null   object
 10  cuota_deposito        3818 non-null   object
 11  cuota_limpieza        3818 non-null   object
 12  precio                3818 non-null   object
dtypes: object(13)
memory usage: 417.6+ KB


In [ ]:
#IMPORTAMOS LA COLUMNA NUMPY PARA TRANSFORMAR LOS TIPOS DE DATOS DE LAS COLUMNAS

import numpy as np

columnas_numericas = ['max_hospedes','cantidad_baños','cantidad_cuartos','cantidad_camas']

datos[columnas_numericas] = datos[columnas_numericas].astype(np.int64)

In [ ]:
#VERIFICAMOS QUE SE CAMBIO EL TIPO DE DATO EN LAS COLUMNAS

datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3818 entries, 0 to 69
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   evaluacion_general    3818 non-null   object
 1   experiencia_local     3818 non-null   object
 2   max_hospedes          3818 non-null   int64 
 3   descripcion_local     3818 non-null   object
 4   descripcion_vecindad  3818 non-null   object
 5   cantidad_baños        3818 non-null   int64 
 6   cantidad_cuartos      3818 non-null   int64 
 7   cantidad_camas        3818 non-null   int64 
 8   modelo_cama           3818 non-null   object
 9   comodidades           3818 non-null   object
 10  cuota_deposito        3818 non-null   object
 11  cuota_limpieza        3818 non-null   object
 12  precio                3818 non-null   object
dtypes: int64(4), object(9)
memory usage: 417.6+ KB


In [ ]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0               10.0                --             1   
0               10.0                --             1   
0               10.0                --             1   
0               10.0                --             1   
0               10.0                --             1   

                                   descripcion_local  \
0  This clean and comfortable one bedroom sits ri...   
0  Our century old Upper Queen Anne house is loca...   
0  Cozy room in two-bedroom apartment along the l...   
0  Very lovely and cozy room for one. Convenientl...   
0  The “Studio at Mibbett Hollow' is in a Beautif...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
0  Upper Queen Anne is a really pleasant, unique ...               1   
0  The convenience of being in Seattle but on the...               1   
0  Ballard is lovely, vibrant and one of the most...               1   
0                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
0                 1               1          Futon   
0                 1               1          Futon   
0                 1               1  Pull-out Sofa   
0                 1               1       Real Bed   

                                         comodidades cuota_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
0  {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
0  {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
0  {"Wireless Internet",Kitchen,"Free Parking on ...             $0   

  cuota_limpieza   precio  
0             $0  $110.00  
0             $0   $45.00  
0             $0   $55.00  
0         $20.00   $52.00  
0         $15.00   $85.00

In [ ]:
datos['evaluacion_general'] = datos['evaluacion_general'].astype(np.float64)

datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3818 entries, 0 to 69
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   evaluacion_general    3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descripcion_local     3818 non-null   object 
 4   descripcion_vecindad  3818 non-null   object 
 5   cantidad_baños        3818 non-null   int64  
 6   cantidad_cuartos      3818 non-null   int64  
 7   cantidad_camas        3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  cuota_deposito        3818 non-null   object 
 11  cuota_limpieza        3818 non-null   object 
 12  precio                3818 non-null   object 
dtypes: float64(1), int64(4), object(8)
memory usage: 417.6+ KB


In [ ]:
#VERIFICAMOS LA COLUMNA precio
#CONTIENE NUMEROS PERO TIENE EL SIMBOLO $
#PARA CONVERTIR EL TIPO DE DATO DE LA COLUMNA ES NECESARIO QUITAR EL SIMBOLO $

datos['precio']

0     $110.00
0      $45.00
0      $55.00
0      $52.00
0      $85.00
       ...   
68    $299.00
68    $199.00
68    $400.00
69    $250.00
69    $350.00
Name: precio, Length: 3818, dtype: object

In [ ]:
#UTILIZAMOS UN FUNCION LAMBDA PARA RECORRER Y ELIMINAR LOS SIMBOLOS Y CARACTERES QUE NO SEAN NUMEROS

datos["precio"] = datos["precio"].apply(lambda x: x.replace("$","").replace(",","").strip())

In [ ]:
#AHORA CAMBIAMOS EL TIPO DE DATO DE LA COLUMNA

datos["precio"] = datos["precio"].astype(np.float64)

In [ ]:
datos["precio"]

0     110.0
0      45.0
0      55.0
0      52.0
0      85.0
      ...  
68    299.0
68    199.0
68    400.0
69    250.0
69    350.0
Name: precio, Length: 3818, dtype: float64

In [ ]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3818 entries, 0 to 69
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   evaluacion_general    3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descripcion_local     3818 non-null   object 
 4   descripcion_vecindad  3818 non-null   object 
 5   cantidad_baños        3818 non-null   int64  
 6   cantidad_cuartos      3818 non-null   int64  
 7   cantidad_camas        3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  cuota_deposito        3818 non-null   object 
 11  cuota_limpieza        3818 non-null   object 
 12  precio                3818 non-null   float64
dtypes: float64(2), int64(4), object(7)
memory usage: 417.6+ KB


In [ ]:
#VERIFICAMOS LAS DEMAS COLUMNAS

datos[["cuota_deposito","cuota_limpieza"]]

cuota_deposito cuota_limpieza
0              $0             $0
0              $0             $0
0              $0             $0
0              $0         $20.00
0              $0         $15.00
..            ...            ...
68      $1,000.00        $178.00
68             $0         $99.00
68             $0             $0
69      $1,000.00        $150.00
69        $500.00             $0

[3818 rows x 2 columns]

In [ ]:
#REALIZAMOS UNA LISTA CON LAS COLUMNAS A MODIFICAR Y UTILIZAMOS LA FUNCION applymap
#LUEGO SE REALIZA LA TRANSFORMACION DE LOS TIPOS DE DATOS DE ESAS COLUMNAS

datos[["cuota_deposito","cuota_limpieza"]] = datos[["cuota_deposito","cuota_limpieza"]].applymap(lambda x: x.replace("$","").replace(",","").strip())
datos[["cuota_deposito","cuota_limpieza"]] = datos[["cuota_deposito","cuota_limpieza"]].astype(np.float64)

<ipython-input-18-8268cb574574>:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  datos[["cuota_deposito","cuota_limpieza"]] = datos[["cuota_deposito","cuota_limpieza"]].applymap(lambda x: x.replace("$","").replace(",","").strip())


In [ ]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3818 entries, 0 to 69
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   evaluacion_general    3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descripcion_local     3818 non-null   object 
 4   descripcion_vecindad  3818 non-null   object 
 5   cantidad_baños        3818 non-null   int64  
 6   cantidad_cuartos      3818 non-null   int64  
 7   cantidad_camas        3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  cuota_deposito        3818 non-null   float64
 11  cuota_limpieza        3818 non-null   float64
 12  precio                3818 non-null   float64
dtypes: float64(4), int64(4), object(5)
memory usage: 417.6+ KB


#**Datos de texto**

In [ ]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   

                                   descripcion_local  \
0  This clean and comfortable one bedroom sits ri...   
0  Our century old Upper Queen Anne house is loca...   
0  Cozy room in two-bedroom apartment along the l...   
0  Very lovely and cozy room for one. Convenientl...   
0  The “Studio at Mibbett Hollow' is in a Beautif...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
0  Upper Queen Anne is a really pleasant, unique ...               1   
0  The convenience of being in Seattle but on the...               1   
0  Ballard is lovely, vibrant and one of the most...               1   
0                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
0                 1               1          Futon   
0                 1               1          Futon   
0                 1               1  Pull-out Sofa   
0                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
0  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
0  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
0  {"Wireless Internet",Kitchen,"Free Parking on ...             0.0   

   cuota_limpieza  precio  
0             0.0   110.0  
0             0.0    45.0  
0             0.0    55.0  
0            20.0    52.0  
0            15.0    85.0

In [ ]:
#CONVERTIMOS LA DESCRIPCION LOCAL EN MINUSCULA PARA COMENZAR A FORMATEARLA PARA LUEGO SER ANALIZADA

datos['descripcion_local'] = datos['descripcion_local'].str.lower()

In [ ]:
datos['descripcion_local']

0     this clean and comfortable one bedroom sits ri...
0     our century old upper queen anne house is loca...
0     cozy room in two-bedroom apartment along the l...
0     very lovely and cozy room for one. convenientl...
0     the “studio at mibbett hollow' is in a beautif...
                            ...                        
68    beautiful craftsman home in the historic wedgw...
68    located in a very easily accessible area of se...
68    this home is fully furnished and available wee...
69    this business-themed modern home features:  *h...
69    this welcoming home is in the quiet residentia...
Name: descripcion_local, Length: 3818, dtype: object

In [ ]:
#ELIMINAR LOS CARACTERES QUE NO UTILICEMOS EN EL ANALISIS DE LA DESCRIPCION
#[^a-zA-Z0-9\-\'] ES PARA INDICAR AL regex LOS CARACTERES QUE NO VA A ELIMINAR
#LAS LETRAS MINUSCULAS DE LA a HASTA LA z, LAS MAYUSCULAS DE LA A HASTA LA Z
#EL GUION Y LA COMILLA SIMPLE

datos['descripcion_local'] = datos['descripcion_local'].str.replace('[^a-zA-Z0-9\-\']',' ', regex=True)

In [ ]:
#ELIMINAMOS LOS GUIONES QUE NO TENGAN NINGUNA PALABRA DEL LADO IZQUIERO Y DERECHO
#(?<!\w)-(?!\w) INDICA QUE EL GUION QUE ESTE EN EL MEDIO DE DOS PALABRAS VA QUEDAR
#SI EL GUION ESTA SIN PALABRAS A LOS LADOS ENTONCES SERA ELIMINADO

datos['descripcion_local'] = datos['descripcion_local'].str.replace('(?<!\w)-(?!\w)',' ', regex=True)

Regex (o expresión regular) es una secuencia de caracteres que define un patrón de búsqueda en un texto. Es una herramienta poderosa y versátil que le permite buscar, reemplazar y manipular patrones de texto de manera eficiente. Regex se usa ampliamente en diferentes áreas, incluida la programación, la ciencia de datos y el procesamiento de textos.

En la ciencia de datos, las expresiones regulares se utilizan a menudo para procesar datos de texto sin formato. Algunas de estas aplicaciones incluyen limpieza de datos, donde se pueden usar expresiones regulares para buscar y reemplazar ciertos caracteres. También se puede utilizar en proyectos de clasificación; un ejemplo es el uso de expresiones regulares para ayudar a encontrar patrones en el texto de los correos electrónicos, lo que ayuda a clasificar si son spam o no.

En general, las expresiones regulares permiten a los científicos de datos procesar, analizar y clasificar grandes volúmenes de datos de texto de forma eficiente y automatizada. El uso adecuado de expresiones regulares puede ayudar a extraer información valiosa de los datos de texto, además de facilitar la limpieza y organización de esos datos.

Puedes crear una expresión regular con la ayuda del sitio web regex101.com. Si quieres saber más sobre regex y su aplicación en bases de datos, vale la pena leer el artículo Principales casos de uso de Regex para procesamiento de datos, que muestra una aplicación regex en banco de datos.

Puede estudiar expresiones regulares y aprender algunas de sus reglas básicas, como las presentadas en el curso, y a medida que se familiarice con el uso de expresiones regulares, podrá comenzar a explorar otras funciones y características más avanzadas para crear patrones más complejos en sus códigos.

In [ ]:
datos['descripcion_local'].head()

0    this clean and comfortable one bedroom sits ri...
0    our century old upper queen anne house is loca...
0    cozy room in two-bedroom apartment along the l...
0    very lovely and cozy room for one  convenientl...
0    the  studio at mibbett hollow' is in a beautif...
Name: descripcion_local, dtype: object

In [ ]:
#CONVERTIR LA DESCRIPCION EN UN UNA LISTA PARA TOKENIZAR LAS PALABRAS
#Y PODER ANALIZAR LA DESCRIPCION Y DETENERMINAR SI ES POSITIVA O NEGATIVA
#SEGUN EL ANALISIS DE LAS PALABRAS QUE CONTENGA

datos['descripcion_local'] = datos['descripcion_local'].str.split()

In [ ]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   

                                   descripcion_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
0  [our, century, old, upper, queen, anne, house,...   
0  [cozy, room, in, two-bedroom, apartment, along...   
0  [very, lovely, and, cozy, room, for, one, conv...   
0  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
0  Upper Queen Anne is a really pleasant, unique ...               1   
0  The convenience of being in Seattle but on the...               1   
0  Ballard is lovely, vibrant and one of the most...               1   
0                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
0                 1               1          Futon   
0                 1               1          Futon   
0                 1               1  Pull-out Sofa   
0                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
0  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
0  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
0  {"Wireless Internet",Kitchen,"Free Parking on ...             0.0   

   cuota_limpieza  precio  
0             0.0   110.0  
0             0.0    45.0  
0             0.0    55.0  
0            20.0    52.0  
0            15.0    85.0

In [ ]:
#AHORA HACER LO MISMO CON LA COLUMNA comodidades
#QUITANDO LOS CARACTERES DE LLAVES, COMILLAS DOBLES
#CON \ INDICAMOS QUE BUSQUE EL CARACTER DE LA LLAVE {
#CON | SE UTILIZA COMO SEPARADOR Y SE LE COLOCA LA OTRA LLAVE
#LUEGO CON LA SIGUIENTE \ SE LE INDICA QUE BUSQUE LAS COMILLAS "

datos['comodidades'] = datos['comodidades'].str.replace('\{|}|\"','', regex=True)

In [ ]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   

                                   descripcion_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
0  [our, century, old, upper, queen, anne, house,...   
0  [cozy, room, in, two-bedroom, apartment, along...   
0  [very, lovely, and, cozy, room, for, one, conv...   
0  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
0  Upper Queen Anne is a really pleasant, unique ...               1   
0  The convenience of being in Seattle but on the...               1   
0  Ballard is lovely, vibrant and one of the most...               1   
0                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
0                 1               1          Futon   
0                 1               1          Futon   
0                 1               1  Pull-out Sofa   
0                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  Internet,Wireless Internet,Kitchen,Free Parkin...             0.0   
0  TV,Internet,Wireless Internet,Kitchen,Free Par...             0.0   
0  TV,Internet,Wireless Internet,Kitchen,Free Par...             0.0   
0  Internet,Wireless Internet,Kitchen,Free Parkin...             0.0   
0  Wireless Internet,Kitchen,Free Parking on Prem...             0.0   

   cuota_limpieza  precio  
0             0.0   110.0  
0             0.0    45.0  
0             0.0    55.0  
0            20.0    52.0  
0            15.0    85.0

In [ ]:
#TRANSFORMAR EN UNA LISTA CON split INDICANDO QUE EL SEPARADOR ES UNA COMA ','

datos['comodidades'] = datos['comodidades'].str.split(',')

In [ ]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   

                                   descripcion_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
0  [our, century, old, upper, queen, anne, house,...   
0  [cozy, room, in, two-bedroom, apartment, along...   
0  [very, lovely, and, cozy, room, for, one, conv...   
0  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
0  Upper Queen Anne is a really pleasant, unique ...               1   
0  The convenience of being in Seattle but on the...               1   
0  Ballard is lovely, vibrant and one of the most...               1   
0                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
0                 1               1          Futon   
0                 1               1          Futon   
0                 1               1  Pull-out Sofa   
0                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
0  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
0  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
0  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
0  [Wireless Internet, Kitchen, Free Parking on P...             0.0   

   cuota_limpieza  precio  
0             0.0   110.0  
0             0.0    45.0  
0             0.0    55.0  
0            20.0    52.0  
0            15.0    85.0

In [ ]:
#AHORA HACER LO MISMO EN LA COLUMNA descripcion_vecindad

# Transformamos el texto en letras minúsculas
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.lower()
# Substituímos los caracteres especiales
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('[^a-zA-Z0-9\-\']', ' ', regex=True)
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('(?<!\w)-(?!\w)', '', regex=True)
# Transformamos el texto en lista, formando el token
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.split()
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   
0                10.0                --             1   

                                   descripcion_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
0  [our, century, old, upper, queen, anne, house,...   
0  [cozy, room, in, two-bedroom, apartment, along...   
0  [very, lovely, and, cozy, room, for, one, conv...   
0  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descripcion_vecindad  cantidad_baños  \
0  [lower, queen, anne, is, near, the, seattle, c...               1   
0  [upper, queen, anne, is, a, really, pleasant, ...               1   
0  [the, convenience, of, being, in, seattle, but...               1   
0  [ballard, is, lovely, vibrant, and, one, of, t...               1   
0                                                 []               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
0                 1               1          Futon   
0                 1               1          Futon   
0                 1               1  Pull-out Sofa   
0                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
0  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
0  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
0  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
0  [Wireless Internet, Kitchen, Free Parking on P...             0.0   

   cuota_limpieza  precio  
0             0.0   110.0  
0             0.0    45.0  
0             0.0    55.0  
0            20.0    52.0  
0            15.0    85.0

##**Desafío: trabajando en otros contextos**
 Siguiente pregunta

Pongamos nuevamente en práctica todo lo que aprendimos durante la clase. He puesto los 2 nuevos conjuntos de datos disponibles para descargar a continuación:

Proyecto Desafío 1: Ventas Online - dados_vendas_clientes.json;
https://cdn3.gnarususercontent.com.br/2928-transformacao-manipulacao-dados/dados_vendas_clientes.json


Proyecto Desafío 2: Administración de Condominios - dados_locacao_imoveis.json.
https://cdn3.gnarususercontent.com.br/2928-transformacao-manipulacao-dados/dados_locacao_imoveis.json

Recuerda: Hay dos proyectos de tratamiento que se construirán durante el curso. Así que guarde su código de construcción para cada desafío para poder aplicarlo a desafíos posteriores.

**Etapa 3**

- **Proyecto Desafío 1: Ventas Online**
En el paso 2, trabajamos en la transformación de datos numéricos. Ahora podemos trabajar con valores textuales.

Debido a una inestabilidad en el sitio web de la empresa, tuvimos problemas con los nombres de los clientes durante el guardado. Esto resultó en una columna de nombres de clientes con una combinación de letras, mayúsculas y minúsculas, números y otros caracteres.

Sabiendo esto, manipula los textos de la columna Cliente para que el resultado sean los nombres de los clientes en letras minúsculas, con ausencia de caracteres especiales o números.

- **Proyecto Desafío 2: Administración de Condominios**
En el paso 2, trabajamos en transformar los datos numéricos. Ahora podemos trabajar con valores textuales.

Buscando explicar la organización de la identificación de los apartamentos, durante la creación del conjunto de datos se añadió el texto (blocoAP). Este texto informa que los nombres de los apartamentos están organizados con la letra mayúscula seguida del número del apartamento. Sin embargo, esto no aporta ninguna información a nuestros datos, por lo que resulta interesante eliminar este texto del conjunto de datos.

Con esto, manipule los textos de la columna apartamento para eliminar el texto (blocoAP) del DataFrame.

##**Proyecto Desafío 1: Ventas Online**

In [ ]:
#IMPORTAR LA LIBRERIA PANDAS
import pandas as pd

url_2 = 'dados_vendas_clientes.json'

datos_2 = pd.read_json(url_2)

#NORMALIZAR EL DATAFRAME

datos_2 = pd.json_normalize(datos_2["dados_vendas"])

#SELECIONAMOS LAS COLUMNAS NECESARIAS QUE CONTIENEN LISTAS DE DATOS PARA EXTRAER

columnas_clientes = list(datos_2.columns)

#EXTRAEMOS LAS COLUMNAS QUE CONTIENEN LISTAS A UTILIZAR

datos_2 = datos_2.explode(columnas_clientes[1:])


datos_2['Cliente'] = datos_2['Cliente'].str.lower()

# Transformamos el texto en letras minúsculas
datos_2['Cliente'] = datos_2['Cliente'].str.replace('[^a-z ]', '', regex=True)


datos_2.head()

Data de venda          Cliente Valor da compra
0    06/06/2022       ana lucia         R$ 836,5
0    06/06/2022  diego armandiu        R$ 573,33
0    06/06/2022  diego armandiu         R$ 392,8
0    06/06/2022  diego armandiu        R$ 512,34
1    07/06/2022  isabely joanes        R$ 825,31

##**Proyecto Desafío 2: Administración de Condominios**

In [ ]:
#IMPORTAR LA LIBRERIA PANDAS
import pandas as pd

url_3 = 'dados_locacao_imoveis.json'

datos_3 = pd.read_json(url_3)

#NORMALIZAR EL DATAFRAME

datos_3 = pd.json_normalize(datos_3["dados_locacao"])

#ELIMINAR (blocoAP) DE LA COLUMNA APARTAMENTOS
#datos_3['apartamento'].str.replace(' \(blocoAP\)', '')

datos_3['apartamento'] = datos_3['apartamento'].str.replace(' (blocoAP)', '')

#columnas_clientes
datos_3.head()

apartamento datas_combinadas_pagamento        datas_de_pagamento  \
0        A101   [01/06/2022, 01/07/2022]  [05/06/2022, 03/07/2022]   
1        A102   [02/06/2022, 02/07/2022]  [02/06/2022, 06/07/2022]   
2        B201   [03/06/2022, 03/07/2022]  [07/06/2022, 03/07/2022]   
3        B202   [04/06/2022, 04/07/2022]  [07/06/2022, 05/07/2022]   
4        C301   [05/06/2022, 05/07/2022]  [10/06/2022, 09/07/2022]   

                      valor_aluguel  
0  [$ 1000,0 reais, $ 2500,0 reais]  
1  [$ 1100,0 reais, $ 2600,0 reais]  
2  [$ 1200,0 reais, $ 2700,0 reais]  
3  [$ 1300,0 reais, $ 2800,0 reais]  
4  [$ 1400,0 reais, $ 2900,0 reais]

#**Datos de tiempo**

In [ ]:
#MOSTRAR DATAFRAME

dt_data = pd.read_json('inmuebles_disponibles.json')

dt_data.head()

id       fecha  lugar_disponible precio
0  857  2016-01-04             False   None
1  857  2016-01-05             False   None
2  857  2016-01-06             False   None
3  857  2016-01-07             False   None
4  857  2016-01-08             False   None

In [ ]:
#CONSULTAR LA INFORMACION DE LAS TABLAS DEL DATAFRAME
dt_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                365000 non-null  int64 
 1   fecha             365000 non-null  object
 2   lugar_disponible  365000 non-null  bool  
 3   precio            270547 non-null  object
dtypes: bool(1), int64(1), object(2)
memory usage: 11.5+ MB


In [ ]:
#MODIFICAR EL TIPO DE DATO DE FECHA

dt_data['fecha'] = pd.to_datetime(dt_data['fecha'])

dt_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                365000 non-null  int64         
 1   fecha             365000 non-null  datetime64[ns]
 2   lugar_disponible  365000 non-null  bool          
 3   precio            270547 non-null  object        
dtypes: bool(1), datetime64[ns](1), int64(1), object(1)
memory usage: 11.5+ MB


###**EJERCICIO: Transformando texto en fechas**



####Usted es responsable de analizar los datos de una empresa que vende productos. Recibiste un conjunto de datos con información sobre las ventas realizadas en los últimos meses, pero notaste que la columna "Fecha de venta" está en formato de cadena (texto), no en formato de fecha. Para realizar el análisis es necesario convertir esta columna al formato de fecha.

Considere el siguiente conjunto de datos con información sobre las ventas de la empresa:

import pandas as pd

## Creando el DataFrame con las informaciones
datos = pd.DataFrame({
    'Fecha de venta': ['01/01/2022', '05/02/2022', '10/03/2022', '15/04/2022','18/04/2022','20/04/2022'],
    'valor': [100, 150, 200, 250,80,180]
})

## Mostrando el DataFrame
print(datos)

## RESPUESTA

datos['Fecha de venta'] = pd.to_datetime(datos['Fecha de venta'], format='%d/%m/%Y')


El formato especificado %d/%m/%Y coincide con el formato original de la fecha en el string. Usando la función pd.to_datetime convertimos la columna “Fecha de venta” y con el parámetro format='%d/%m/%Y' podemos indicar el formato correcto del string, que es "día/mes/año".

In [ ]:
#MOSTRAR SOLO EL AÑO Y EL MES EN LA COLUMNA fecha

dt_data['fecha'].dt.strftime('%Y-%m')


0         2016-01
1         2016-01
2         2016-01
3         2016-01
4         2016-01
           ...   
364995    2016-12
364996    2016-12
364997    2016-12
364998    2017-01
364999    2017-01
Name: fecha, Length: 365000, dtype: object

In [ ]:
#AGRUPAR POR AÑO Y MES PARA CONTAR LOS REGISTROS QUE CONTIENEN True EN LA COLUMNA lugar_disponible

dt_data.groupby(dt_data['fecha'].dt.strftime('%Y-%m'))['lugar_disponible'].sum()

fecha
2016-01    16543
2016-02    20128
2016-03    23357
2016-04    22597
2016-05    23842
2016-06    23651
2016-07    22329
2016-08    22529
2016-09    22471
2016-10    23765
2016-11    23352
2016-12    24409
2017-01     1574
Name: lugar_disponible, dtype: int64

##**06 Desafío: hazlo tú mismo**
 Siguiente pregunta

En esta clase, aprendimos cómo manipular datos temporales usando datetime. Entendimos cómo transformar una columna a fecha y hora y luego manipular estos datos. Aun así, no todos los datos del conjunto inmuebles_disponibles.json fueron tratados.

Durante las clases de este curso, descubrimos cómo transformar y trabajar con valores numéricos, por ejemplo, eliminando valores numéricos dentro de un texto y transformándolos en un tipo numérico, como int64 o float64.

Sabiendo esto, en esta actividad te propongo transformar los datos de la columna precio del conjunto de datos inmuebles_disponibles.json al tipo numérico float64. Recordando que, antes de hacer esto, debes llenar los valores vacíos de la columna con un valor. Una sugerencia: reemplazar con el string '0.0'.

No dudes en seguir los mismos pasos dados en clase o, si lo prefieres, realizar otras mejoras, como eliminar algunos caracteres o palabras vacías. En el apartado “Opinión del instructor” encontrarás una posible resolución para esta actividad.

In [ ]:
dt_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                365000 non-null  int64         
 1   fecha             365000 non-null  datetime64[ns]
 2   lugar_disponible  365000 non-null  bool          
 3   precio            270547 non-null  object        
dtypes: bool(1), datetime64[ns](1), int64(1), object(1)
memory usage: 11.5+ MB


In [ ]:
dt_data['precio']

0            None
1            None
2            None
3            None
4            None
           ...   
364995    $140.00
364996    $140.00
364997    $140.00
364998    $140.00
364999       None
Name: precio, Length: 365000, dtype: object

In [ ]:
# utilizamos el método fillna para llenar los elementos vacíos por '0.0'
# definimos el parámetro de inplace para True para substituir en el DataFrame
dt_data['precio'].fillna('0.0', inplace = True)
dt_data['precio']

0             0.0
1             0.0
2             0.0
3             0.0
4             0.0
           ...   
364995    $140.00
364996    $140.00
364997    $140.00
364998    $140.00
364999        0.0
Name: precio, Length: 365000, dtype: object

In [ ]:
#ELIMINAMOS LOS CARACTERES $

dt_data['precio'] = dt_data['precio'].apply(lambda x: x.replace('$','').replace(',',''))

In [ ]:
#CONVERTIMOS EL TIPO DE DATO DE LA COLUMNA A FLOAT

dt_data['precio'] = dt_data['precio'].astype(np.float64)

In [ ]:
dt_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                365000 non-null  int64         
 1   fecha             365000 non-null  datetime64[ns]
 2   lugar_disponible  365000 non-null  bool          
 3   precio            365000 non-null  float64       
dtypes: bool(1), datetime64[ns](1), float64(1), int64(1)
memory usage: 11.5 MB


#**07 Desafío: trabajando en otros contextos**
 - **Siguiente pregunta**

Pongamos nuevamente en práctica todo lo que aprendimos durante la clase. He puesto los 2 nuevos conjuntos de datos disponibles para descargar a continuación:

Proyecto Desafío 1: Ventas Online - dados_vendas_clientes.json;

Proyecto Desafío 2: Administración de Condominios - dados_locacao_imoveis.json.

Recuerda: Estos dos proyectos de tratamiento se construyeron durante el curso. Por lo tanto, considere desarrollos previos para realizar este paso final.

Etapa 4

- **Proyecto Desafío 1: Ventas Online**
En los pasos anteriores ya hemos trabajado con varios tipos de datos, ahora podemos trabajar con datos de tiempo.

En la columna Fecha de venta tenemos fechas en el formato 'día/mes/año' (dd/mm/AAAA). Transforme estos datos al tipo datetime y busque una forma de visualización de subconjunto que pueda contribuir al objetivo del contexto en el que se insertan los datos.

Si no recuerdas el problema del Proyecto Desafío 1, te dejo el texto de la situación a continuación para que sea más fácil encontrar la información:

El objetivo de este proyecto es analizar los resultados de un evento con los clientes de una empresa de venta online. Se recopiló un conjunto de datos que contiene los clientes que gastaron más en productos dentro de los 5 días posteriores a la venta, que es la duración del evento. Este análisis identificará al cliente con la mayor compra esa semana, quien recibirá un premio de la tienda, y posteriormente, puede ayudar a la empresa a crear nuevas estrategias para atraer más clientes.


- **Proyecto Desafío 2: Administración de Condominios**
Al igual que en el Proyecto Desafío 1, trabajamos con todas las columnas excepto las que involucran fechas.

En las columnas datas_de_pagamento y datas_combinadas_pagamento tenemos fechas en el formato 'día/mes/año' (dd/mm/AAAA). Transforme estos datos al tipo datetime y busque una forma de visualización de subconjunto que pueda contribuir al objetivo del contexto en el que se insertan los datos.

Si no recuerdas el problema del Proyecto Desafío 2, te dejo el texto de la situación a continuación para que sea más fácil encontrar la información:

Administrar condominios es una tarea que requiere mucha atención y organización. Entre las diversas responsabilidades de gestión se encuentra el cobro del alquiler a los inquilinos. Para garantizar la buena salud financiera de la empresa, es fundamental que estos pagos se realicen de forma regular y puntual. Sin embargo, sabemos que esto no siempre sucede. Teniendo esto en cuenta, propongo un desafío de procesamiento de datos con el objetivo de analizar el retraso en el pago de la renta en el condominio ficticio de algunos residentes.



###**Proyecto Desafío 1: Ventas Online**

In [ ]:
datos_2.sample(10)

Data de venda          Cliente Valor da compra
1    07/06/2022  isabely joanes        R$ 314,69
4    10/06/2022    paulo castro        R$ 899,16
4    10/06/2022   thiago fritzz        R$ 885,24
0    06/06/2022  diego armandiu        R$ 573,33
0    06/06/2022  diego armandiu        R$ 512,34
3    09/06/2022  julya meireles         R$ 390,3
1    07/06/2022  isabely joanes        R$ 339,18
1    07/06/2022  isabely joanes        R$ 168,07
4    10/06/2022     pedro pasco        R$ 311,15
2    08/06/2022     joo gabriel        R$ 386,34

In [ ]:
datos_2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 20 entries, 0 to 4
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Data de venda    20 non-null     object
 1   Cliente          20 non-null     object
 2   Valor da compra  20 non-null     object
dtypes: object(3)
memory usage: 1.2+ KB


In [ ]:
#CONVERTIR EL TIPO DE DATO DE LA COLUMNA FECHA
datos_2['Data de venda'] = pd.to_datetime(datos_2['Data de venda'],format='%d/%m/%Y')

In [ ]:
datos_2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 20 entries, 0 to 4
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Data de venda    20 non-null     datetime64[ns]
 1   Cliente          20 non-null     object        
 2   Valor da compra  20 non-null     object        
dtypes: datetime64[ns](1), object(2)
memory usage: 1.2+ KB


In [ ]:
#ELIMINAR CARACTERES INNECESARIOS EN LA COLUMNA Valor da compra PARA PODER EJECUTAR OPERACIONES

datos_2["Valor da compra"] = datos_2["Valor da compra"].apply(lambda x: x.replace("R$ ","").replace(",",".").strip())

In [ ]:
datos_2["Valor da compra"]

0     836.5
0    573.33
0     392.8
0    512.34
1    825.31
1    168.07
1    339.18
1    314.69
2    682.05
2    386.34
2    622.65
2    630.79
3     390.3
3    759.16
3    334.47
3    678.78
4    314.24
4    311.15
4    899.16
4    885.24
Name: Valor da compra, dtype: object

In [ ]:
#CONVERTIMOS EL TIPO DE DATO DE LA COLUMNA A FLOAT

datos_2['Valor da compra'] = datos_2['Valor da compra'].astype(np.float64)

In [ ]:
datos_2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 20 entries, 0 to 4
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Data de venda    20 non-null     datetime64[ns]
 1   Cliente          20 non-null     object        
 2   Valor da compra  20 non-null     float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 1.2+ KB


In [ ]:
# Calcular el total recaudado en compras por cada cliente
total_compras = datos_2.groupby(['Cliente'])['Valor da compra'].sum()

In [ ]:
total_compras

Cliente
ana lucia           836.50
diego armandiu     1478.47
isabely joanes     2329.30
joo gabriel         386.34
julya meireles     1643.74
maria julia        2086.65
paulo castro        899.16
pedro pasco         311.15
thiago fritzz       885.24
Name: Valor da compra, dtype: float64

###**Proyecto Desafío 2: Administración de Condominios**

In [ ]:
datos_3.sample(10)

apartamento datas_combinadas_pagamento datas_de_pagamento   valor_aluguel  \
9         E502                 2022-07-10         2022-07-12  $ 1600,0 reais   
3         B202                 2022-06-04         2022-06-07  $ 1300,0 reais   
11        F602                 2022-06-12         2022-06-15  $ 2100,0 reais   
10        F601                 2022-07-11         2022-07-13  $ 1000,0 reais   
5         C302                 2022-06-06         2022-06-08  $ 1500,0 reais   
14        H801                 2022-07-15         2022-07-16  $ 2200,0 reais   
13        G702                 2022-07-14         2022-07-17  $ 2100,0 reais   
7         D402                 2022-07-08         2022-07-14  $ 1400,0 reais   
0         A101                 2022-06-01         2022-06-05  $ 1000,0 reais   
2         B201                 2022-07-03         2022-07-03  $ 2700,0 reais   

    atraso  
9        2  
3        3  
11       3  
10       2  
5        2  
14       1  
13       3  
7        6  
0        4  
2        0

In [ ]:
#SELECIONAMOS LAS COLUMNAS NECESARIAS QUE CONTIENEN LISTAS DE DATOS PARA EXTRAER

columnas_fechas = list(datos_3.columns)
columnas_fechas

['apartamento',
 'datas_combinadas_pagamento',
 'datas_de_pagamento',
 'valor_aluguel',
 'atraso']

In [ ]:
#EXTRAEMOS LAS COLUMNAS QUE CONTIENEN LISTAS A UTILIZAR

datos_3 = datos_3.explode(columnas_fechas[1:])
datos_3

apartamento datas_combinadas_pagamento datas_de_pagamento   valor_aluguel  \
0         A101                 2022-06-01         2022-06-05  $ 1000,0 reais   
0         A101                 2022-07-01         2022-07-03  $ 2500,0 reais   
1         A102                 2022-06-02         2022-06-02  $ 1100,0 reais   
1         A102                 2022-07-02         2022-07-06  $ 2600,0 reais   
2         B201                 2022-06-03         2022-06-07  $ 1200,0 reais   
2         B201                 2022-07-03         2022-07-03  $ 2700,0 reais   
3         B202                 2022-06-04         2022-06-07  $ 1300,0 reais   
3         B202                 2022-07-04         2022-07-05  $ 2800,0 reais   
4         C301                 2022-06-05         2022-06-10  $ 1400,0 reais   
4         C301                 2022-07-05         2022-07-09  $ 2900,0 reais   
5         C302                 2022-06-06         2022-06-08  $ 1500,0 reais   
5         C302                 2022-07-06         2022-07-12  $ 1200,0 reais   
6         D401                 2022-06-07         2022-06-07  $ 1600,0 reais   
6         D401                 2022-07-07         2022-07-09  $ 1300,0 reais   
7         D402                 2022-06-08         2022-06-10  $ 1700,0 reais   
7         D402                 2022-07-08         2022-07-14  $ 1400,0 reais   
8         E501                 2022-06-09         2022-06-10  $ 1800,0 reais   
8         E501                 2022-07-09         2022-07-09  $ 1500,0 reais   
9         E502                 2022-06-10         2022-06-16  $ 1900,0 reais   
9         E502                 2022-07-10         2022-07-12  $ 1600,0 reais   
10        F601                 2022-06-11         2022-06-17  $ 2000,0 reais   
10        F601                 2022-07-11         2022-07-13  $ 1000,0 reais   
11        F602                 2022-06-12         2022-06-15  $ 2100,0 reais   
11        F602                 2022-07-12         2022-07-12  $ 1100,0 reais   
12        G701                 2022-06-13         2022-06-19  $ 2200,0 reais   
12        G701                 2022-07-13         2022-07-20  $ 1200,0 reais   
13        G702                 2022-06-14         2022-06-15  $ 2300,0 reais   
13        G702                 2022-07-14         2022-07-17  $ 2100,0 reais   
14        H801                 2022-06-15         2022-06-18  $ 2400,0 reais   
14        H801                 2022-07-15         2022-07-16  $ 2200,0 reais   

    atraso  
0        4  
0        2  
1        0  
1        4  
2        4  
2        0  
3        3  
3        1  
4        5  
4        4  
5        2  
5        6  
6        0  
6        2  
7        2  
7        6  
8        1  
8        0  
9        6  
9        2  
10       6  
10       2  
11       3  
11       0  
12       6  
12       7  
13       1  
13       3  
14       3  
14       1

In [ ]:
datos_3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30 entries, 0 to 14
Data columns (total 4 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   apartamento                 30 non-null     object
 1   datas_combinadas_pagamento  30 non-null     object
 2   datas_de_pagamento          30 non-null     object
 3   valor_aluguel               30 non-null     object
dtypes: object(4)
memory usage: 1.2+ KB


In [ ]:
#MODIFICAR EL TIPO DE DATO DE FECHA

datos_3['datas_combinadas_pagamento'] = pd.to_datetime(datos_3['datas_combinadas_pagamento'], format='%d/%m/%Y')
datos_3['datas_de_pagamento'] = pd.to_datetime(datos_3['datas_de_pagamento'], format='%d/%m/%Y')
datos_3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30 entries, 0 to 14
Data columns (total 4 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   apartamento                 30 non-null     object        
 1   datas_combinadas_pagamento  30 non-null     datetime64[ns]
 2   datas_de_pagamento          30 non-null     datetime64[ns]
 3   valor_aluguel               30 non-null     object        
dtypes: datetime64[ns](2), object(2)
memory usage: 1.2+ KB


In [ ]:
#CALCULAMOS LOS DIAS DE ATRASO EN LOS PAGOS RESTANDO LA FECHA datas_de_pagamento DE LA FECHA datas_combinadas_pagamento
datos_3['atraso'] = (datos_3['datas_de_pagamento'] - datos_3['datas_combinadas_pagamento']).dt.days


In [ ]:
datos_3

apartamento datas_combinadas_pagamento datas_de_pagamento   valor_aluguel  \
0         A101                 2022-06-01         2022-06-05  $ 1000,0 reais   
0         A101                 2022-07-01         2022-07-03  $ 2500,0 reais   
1         A102                 2022-06-02         2022-06-02  $ 1100,0 reais   
1         A102                 2022-07-02         2022-07-06  $ 2600,0 reais   
2         B201                 2022-06-03         2022-06-07  $ 1200,0 reais   
2         B201                 2022-07-03         2022-07-03  $ 2700,0 reais   
3         B202                 2022-06-04         2022-06-07  $ 1300,0 reais   
3         B202                 2022-07-04         2022-07-05  $ 2800,0 reais   
4         C301                 2022-06-05         2022-06-10  $ 1400,0 reais   
4         C301                 2022-07-05         2022-07-09  $ 2900,0 reais   
5         C302                 2022-06-06         2022-06-08  $ 1500,0 reais   
5         C302                 2022-07-06         2022-07-12  $ 1200,0 reais   
6         D401                 2022-06-07         2022-06-07  $ 1600,0 reais   
6         D401                 2022-07-07         2022-07-09  $ 1300,0 reais   
7         D402                 2022-06-08         2022-06-10  $ 1700,0 reais   
7         D402                 2022-07-08         2022-07-14  $ 1400,0 reais   
8         E501                 2022-06-09         2022-06-10  $ 1800,0 reais   
8         E501                 2022-07-09         2022-07-09  $ 1500,0 reais   
9         E502                 2022-06-10         2022-06-16  $ 1900,0 reais   
9         E502                 2022-07-10         2022-07-12  $ 1600,0 reais   
10        F601                 2022-06-11         2022-06-17  $ 2000,0 reais   
10        F601                 2022-07-11         2022-07-13  $ 1000,0 reais   
11        F602                 2022-06-12         2022-06-15  $ 2100,0 reais   
11        F602                 2022-07-12         2022-07-12  $ 1100,0 reais   
12        G701                 2022-06-13         2022-06-19  $ 2200,0 reais   
12        G701                 2022-07-13         2022-07-20  $ 1200,0 reais   
13        G702                 2022-06-14         2022-06-15  $ 2300,0 reais   
13        G702                 2022-07-14         2022-07-17  $ 2100,0 reais   
14        H801                 2022-06-15         2022-06-18  $ 2400,0 reais   
14        H801                 2022-07-15         2022-07-16  $ 2200,0 reais   

    atraso  
0        4  
0        2  
1        0  
1        4  
2        4  
2        0  
3        3  
3        1  
4        5  
4        4  
5        2  
5        6  
6        0  
6        2  
7        2  
7        6  
8        1  
8        0  
9        6  
9        2  
10       6  
10       2  
11       3  
11       0  
12       6  
12       7  
13       1  
13       3  
14       3  
14       1

In [ ]:
# Calcular el promedio de tiempo de atraso por apartamentos
media_atraso = datos_3.groupby(['apartamento'])['atraso'].mean()

# Visualizar el resultado
media_atraso

apartamento
A101    3.0
A102    2.0
B201    2.0
B202    2.0
C301    4.5
C302    4.0
D401    1.0
D402    4.0
E501    0.5
E502    4.0
F601    4.0
F602    1.5
G701    6.5
G702    2.0
H801    2.0
Name: atraso, dtype: float64